### PR TITLE
Delete image

### LINKED ISSUE #8: Delete Image
#### As a user I want to delete an image.

__Story__:
Our users are awesome at adding images to our web service. Some of our users
add dozens in a single session. Occasionally, however, some users add the wrong
image, and would like a way to remove those images. Please provide our users
with the ability to delete images. Note that at this point we trust our users
not to abuse this feature and delete more images than they should.

__Acceptance criteria__:
- [ ] The user is prompted for confirmation prior to deletion. (Use javascript
  confirm() dialog)
- [ ] Deleted images are removed from the homepage.
- [ ] A flow test using page objects exists for the flow to create and
  subsequently delete an image.

__Dependencies__:
- Image Index

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm
- https://github.com/appfolio/ae_page_objects
- http://martinfowler.com/bliki/PageObject.html
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects
- https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings#javascript-dialogs

__Flow Test Info__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the [ae_page_objects readme](https://github.com/appfolio/ae_page_objects)
which will be an extremely helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Important Notes__:

Upgrade your Firefox to the latest version.

When dealing with javascript confirm() dialog in the flow test, this [reference](https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings#javascript-dialogs)
will be helpful.

### PR DESCRIPTION
**Pull request by gauravc88**
Closes #8 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/page_objects/images/new_page.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    class NewPage < PageObjects::Document
+      path :new_image
+      path :images # from POST to images for failed create
+
+      form_for :image do
+        element :url
+        element :tag_list
+      end
+
+      def create_image!(url: nil, tags: nil)
```
Comment (brightmer) [inline test/page_objects/images/new_page.rb]: just fyi: `def create_image!(url:, tags:)` is equivalent

Comment (gauravc88) [reply test/page_objects/images/new_page.rb]: ah nice, good to know

File: test/flows/images_crud_test.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    images_index_page = image_show_page.delete_and_confirm!
+    assert_equal "Image deleted\n×", images_index_page.flash_message(:notice)
+
+    assert_equal 1, images_index_page.images.count
+    assert_not images_index_page.showing_image?(url: ugly_cat_url)
+    assert images_index_page.showing_image?(url: cute_puppy_url)
+  end
+
+  test 'view images associated with a tag' do
```
[the remark this answers is not in the document]
Comment (gauravc88) [reply test/flows/images_crud_test.rb]: I actually need to change the def to test syntax. The def syntax was supposed to be temporary for debugging, would change it back to use the 'test' syntax

File: test/flows/images_crud_test.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    ).as_a(PageObjects::Images::NewPage)
+    assert_equal 'Url must start with http or https and Url must have a valid extension',
+                 new_image_page.url.error_message
+
+    image_url = 'https://media3.giphy.com/media/EldfH1VJdbrwY/200.gif'
+    new_image_page.url.set(image_url)
+    new_image_page.tag_list.set(tags.join(','))
+
+    image_show_page = new_image_page.create_image!
+    assert_equal "Image saved\n×", image_show_page.flash_message(:notice)
```
[the remark this answers is not in the document]
Comment (gauravc88) [reply test/flows/images_crud_test.rb]: Yes, it's for the close button in the flash message.

File: test/flow_test_helper.rb [hunk 4]
```diff
@@ -3,6 +3,14 @@
 require 'active_support/test_case'
 require 'capybara/rails'
 require 'capybara/dsl'
+require 'ae_page_objects'
+require 'ae_page_objects/rails'
+
+module PageObjects
```
[the remark this answers is not in the document]
Comment (gauravc88) [reply test/flow_test_helper.rb]: This file was included when i merged from the flowTest branch. So not sure what the original author intended, I am guessing it just declares an empty PageObjects module to be extended later

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -137,6 +138,22 @@ def test_show
...
+    assert_redirected_to images_path
+    assert_equal 'Image deleted', flash[:notice]
+  end
+
+  def test_destroy__non_existent_image
+    assert_difference('Image.count', 0) do
+      delete image_path(-1)
+    end
+    assert_redirected_to images_path
+    assert_equal 'Image deleted', flash[:notice]
```
[the remark this answers is not in the document]
Comment (gauravc88) [reply test/controllers/images_controller_test.rb]: So the delete logic notifies the user that the image was deleted even if it wasn't found. The primary case was if the same image was opened on two different tabs and deleted on one of them. After this if the user tries to delete the image from the second tab, the user still sees the image as deleted even though its record isn't found. The approach was recommended by Bryce who was pair programming with me.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -25,6 +25,16 @@ def show
...
   end
 
+  def destroy
+    begin
+      Image.destroy(params[:id])
+    rescue ActiveRecord::RecordNotFound
+      1
+    end
+    flash[:notice] = 'Image deleted'
+    redirect_to images_path
```
[the remark this answers is not in the document]
Comment (gauravc88) [reply app/controllers/images_controller.rb]: So the delete logic notifies the user that the image was deleted even if it wasn't found. The primary case was if the same image was opened on two different tabs and deleted on one of them. After this if the user tries to delete the image from the second tab, the user still sees the image as deleted even though its record isn't found. The approach was recommended by Bryce who was pair programming with me.

**Pull request merged**